java.lang.IllegalArgumentException: wrong number of arguments

### TestNG Version
> Note: only the latest version is supported

Regression from 6.11 to 6.14.2

```
The factory method class com.polo.primetest.factories.FactoryTest.createInstances() threw an exception
    at org.testng.internal.FactoryMethod.invoke(FactoryMethod.java:197)
    at org.testng.internal.TestNGClassFinder.processFactory(TestNGClassFinder.java:223)
    at org.testng.internal.TestNGClassFinder.processMethod(TestNGClassFinder.java:179)
    at org.testng.internal.TestNGClassFinder.processClass(TestNGClassFinder.java:171)
    at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:121)
    at org.testng.TestRunner.initMethods(TestRunner.java:370)
    at org.testng.TestRunner.init(TestRunner.java:271)
    at org.testng.TestRunner.init(TestRunner.java:241)
    at org.testng.TestRunner.<init>(TestRunner.java:167)
    at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:663)
    at org.testng.SuiteRunner.init(SuiteRunner.java:260)
    at org.testng.SuiteRunner.<init>(SuiteRunner.java:198)
    at org.testng.TestNG.createSuiteRunner(TestNG.java:1295)
    at org.testng.TestNG.createSuiteRunners(TestNG.java:1273)
    at org.testng.TestNG.runSuitesLocally(TestNG.java:1128)
    at org.testng.TestNG.runSuites(TestNG.java:1049)
    at org.testng.TestNG.run(TestNG.java:1017)
    at com.polo.primetest.testsupport.TestInitiator.createTestNG(TestInitiator.java:436)
    at com.polo.primetest.testsupport.TestInitiator.init(TestInitiator.java:302)
    at com.polo.primetest.execution.main.Main.main(Main.java:22)
Caused by: java.lang.IllegalArgumentException: wrong number of arguments
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.testng.internal.FactoryMethod.invoke(FactoryMethod.java:167)
    ... 19 common frames omitted
```

```java
public class FactoryTest
    {
        @Factory
        public Object[] createInstances( ITestContext ctx ) throws Exception
        {
            XmlTest currentXmlTest = ctx.getCurrentXmlTest();
            String parameterValue = currentXmlTest.getParameter( currentXmlTest.getName() );
        }
```

See https://stackoverflow.com/q/49570901/4234729

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-1745: Support native injection for @Factory methods (Krishnan Mahadevan)
 Fixed: GITHUB-1746: Make InvokedMethodNameListener thread-safe, fixing occasional build failures (Sergei Tachenov)
 Fixed: GITHUB-1538: Dependent methods don't get invoked when a failed test method is retried via a RetryAnalyser (Krishnan Mahadevan)
 Fixed: GITHUB-426: firstTimeOnly ignored for threadPoolSize > 1 (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/FactoryMethod.java
+++ b/src/main/java/org/testng/internal/FactoryMethod.java
@@ -155,7 +155,7 @@ public class FactoryMethod extends BaseTestMethod {
             m_testContext.getCurrentXmlTest().getSuite(),
             m_annotationFinder,
             null /* fedInstance */,
-            m_dataProviderListeners.values()).parameters;
+            m_dataProviderListeners.values(), "@Factory").parameters;
 
     try {
       List<Integer> indices = factoryAnnotation.getIndices();

--- a/src/main/java/org/testng/internal/Parameters.java
+++ b/src/main/java/org/testng/internal/Parameters.java
@@ -85,7 +85,8 @@ public class Parameters {
       BeforeGroups.class,
       AfterGroups.class,
       BeforeMethod.class,
-      AfterMethod.class
+      AfterMethod.class,
+      Factory.class
   };
 
   private static Map<String, List<Class<?>>> mapping = Maps.newHashMap();
@@ -115,6 +116,8 @@ public class Parameters {
           +--------------+--------------+---------+--------+----------+-------------+
           | Test         | Yes          | Yes     | Yes    | No       | No          |
           +--------------+--------------+---------+--------+----------+-------------+
+          | Factory      | Yes          | Yes     | No     | No       | No          |
+          +--------------+--------------+---------+--------+----------+-------------+
 
  */
   static {
@@ -135,7 +138,8 @@ public class Parameters {
 
     mapping.put(BeforeMethod.class.getSimpleName(), beforeAfterMethod);
     mapping.put(AfterMethod.class.getSimpleName(), beforeAfterMethod);
-    mapping.put(Test.class.getSimpleName(), Arrays.<Class<?>>asList(ITestContext.class, XmlTest.class, Method.class));
+    mapping.put(Test.class.getSimpleName(), Arrays.asList(ITestContext.class, XmlTest.class, Method.class));
+    mapping.put(Factory.class.getSimpleName(), ctxTest);
 
   }
 
@@ -698,14 +702,24 @@ public class Parameters {
    * @return An Iterator over the values for each parameter of this
    * method.
    */
-  public static ParameterHolder handleParameters(final ITestNGMethod testMethod,
-                                                 final Map<String, String> allParameterNames,
-                                                 final Object instance,
-                                                 final MethodParameters methodParams,
-                                                 final XmlSuite xmlSuite,
-                                                 final IAnnotationFinder annotationFinder,
-                                                 final Object fedInstance,
-                                                 final Collection<IDataProviderListener> dataProviderListeners) {
+  public static ParameterHolder handleParameters(ITestNGMethod testMethod, Map<String, String> allParameterNames,
+                                                 Object instance, MethodParameters methodParams, XmlSuite xmlSuite,
+                                                 IAnnotationFinder annotationFinder, Object fedInstance,
+                                                 Collection<IDataProviderListener> dataProviderListeners) {
+    return handleParameters(testMethod, allParameterNames, instance, methodParams, xmlSuite, annotationFinder,
+            fedInstance, dataProviderListeners, "@Test");
+  }
+
+  /**
+   * If the method has parameters, fill them in. Either by using a @DataProvider
+   * if any was provided, or by looking up <parameters> in testng.xml
+   * @return An Iterator over the values for each parameter of this
+   * method.
+   */
+  public static ParameterHolder handleParameters(ITestNGMethod testMethod, Map<String, String> allParameterNames,
+                                                 Object instance, MethodParameters methodParams, XmlSuite xmlSuite,
+                                                 IAnnotationFinder annotationFinder, Object fedInstance,
+                                                 Collection<IDataProviderListener> dataProviderListeners, String annotationName) {
     /*
      * Do we have a @DataProvider? If yes, then we have several
      * sets of parameters for this method
@@ -788,23 +802,23 @@ public class Parameters {
     else {
       origin = ParameterOrigin.ORIGIN_XML;
     }
-      //
-      // Normal case: we have only one set of parameters coming from testng.xml
-      //
-      allParameterNames.putAll(methodParams.xmlParameters);
-      // Create an Object[][] containing just one row of parameters
-      Object[][] allParameterValuesArray = new Object[1][];
-      allParameterValuesArray[0] = createParameters(testMethod.getConstructorOrMethod().getMethod(),
-              methodParams, annotationFinder, xmlSuite, ITestAnnotation.class, "@Test");
+    //
+    // Normal case: we have only one set of parameters coming from testng.xml
+    //
+    allParameterNames.putAll(methodParams.xmlParameters);
+    // Create an Object[][] containing just one row of parameters
+    Object[][] allParameterValuesArray = new Object[1][];
+    allParameterValuesArray[0] = createParameters(testMethod.getConstructorOrMethod().getMethod(),
+            methodParams, annotationFinder, xmlSuite, ITestAnnotation.class, annotationName);
 
-      // Mark that this method needs to have at least a certain
-      // number of invocations (needed later to call AfterGroups
-      // at the right time).
-      testMethod.setParameterInvocationCount(allParameterValuesArray.length);
-      // Turn it into an Iterable
-      Iterator<Object[]> parameters = new ArrayIterator(allParameterValuesArray);
+    // Mark that this method needs to have at least a certain
+    // number of invocations (needed later to call AfterGroups
+    // at the right time).
+    testMethod.setParameterInvocationCount(allParameterValuesArray.length);
+    // Turn it into an Iterable
+    Iterator<Object[]> parameters = new ArrayIterator(allParameterValuesArray);
 
-      return new ParameterHolder(parameters, origin, null);
+    return new ParameterHolder(parameters, origin, null);
   }
 
   /**

--- a/src/test/java/test/factory/issue1745/Github1745Test.java
+++ b/src/test/java/test/factory/issue1745/Github1745Test.java
@@ -1,0 +1,29 @@
+package test.factory.issue1745;
+
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlTest;
+import test.InvokedMethodNameListener;
+import test.SimpleBaseTest;
+
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class Github1745Test extends SimpleBaseTest {
+
+    private static final String TEST_METHOD = "testMethod";
+
+    @Test
+    public void testMethod() {
+        XmlTest xmlTest = createXmlTest("1745_suite", "1745_test", SuiteXmlPoweredFactoryTest.class);
+        xmlTest.addParameter("number", "3");
+        TestNG testng = create();
+        testng.setXmlSuites(Collections.singletonList(xmlTest.getSuite()));
+        InvokedMethodNameListener listener = new InvokedMethodNameListener();
+        testng.addListener(listener);
+        testng.run();
+        assertThat(listener.getSucceedMethodNames())
+                .containsExactly(TEST_METHOD, TEST_METHOD, TEST_METHOD, TEST_METHOD);
+    }
+}

--- a/src/test/java/test/factory/issue1745/SuiteXmlPoweredFactoryTest.java
+++ b/src/test/java/test/factory/issue1745/SuiteXmlPoweredFactoryTest.java
@@ -1,0 +1,29 @@
+package test.factory.issue1745;
+
+import org.testng.ITestContext;
+import org.testng.annotations.Factory;
+import org.testng.xml.XmlTest;
+
+public class SuiteXmlPoweredFactoryTest {
+
+    @Factory
+    public Object[] createInstances(ITestContext ctx) {
+        return new Object[]{new TestClassSample(ctx, "testcontext")};
+    }
+
+    @Factory
+    public Object[] createInstances(XmlTest currentXmlTest) {
+        return new Object[]{new TestClassSample(currentXmlTest, "xmltest")};
+    }
+
+    @Factory
+    public Object[] createInstances(ITestContext ctx, XmlTest currentXmlTest) {
+        return new Object[]{new TestClassSample(ctx, currentXmlTest)};
+    }
+
+    @Factory
+    public Object[] createInstances(XmlTest currentXmlTest, ITestContext ctx) {
+        return new Object[]{new TestClassSample(ctx, currentXmlTest)};
+    }
+
+}

--- a/src/test/java/test/factory/issue1745/TestClassSample.java
+++ b/src/test/java/test/factory/issue1745/TestClassSample.java
@@ -1,0 +1,40 @@
+package test.factory.issue1745;
+
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+
+import org.testng.ITestContext;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlTest;
+
+public class TestClassSample {
+    private ITestContext ctx;
+    private XmlTest currentXmlTest;
+    private String whatWasPassed = "";
+
+    public TestClassSample(ITestContext ctx, String whatWasPassed) {
+        this.ctx = ctx;
+        this.whatWasPassed = whatWasPassed;
+    }
+
+    public TestClassSample(XmlTest currentXmlTest, String whatWasPassed) {
+        this.currentXmlTest = currentXmlTest;
+        this.whatWasPassed = whatWasPassed;
+    }
+
+    public TestClassSample(ITestContext ctx, XmlTest currentXmlTest) {
+        this.ctx = ctx;
+        this.currentXmlTest = currentXmlTest;
+    }
+
+    @Test
+    public void testMethod() {
+        if ("testcontext".equalsIgnoreCase(whatWasPassed)) { //only testcontext was provided
+            assertNotNull(ctx);
+            currentXmlTest = ctx.getCurrentXmlTest();
+        }
+        assertNotNull(currentXmlTest);
+        int number = Integer.parseInt(currentXmlTest.getParameter("number"));
+        assertTrue(number != 0);
+    }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -174,6 +174,7 @@
       <class name="test.github1461.MemoryLeakTestNg" />
       <!-- TODO fix the random issue <class name="test.issue565.Issue565Test"/>-->
       <class name="test.factory.github328.GitHub328Test" />
+      <class name="test.factory.issue1745.Github1745Test"/>
       <class name="test.github1490.VerifyDataProviderListener"/>      
       <class name="test.methodselection.MethodSelectionTest"/>
       <class name="test.beforegroups.BeforeGroupsTest"/>

